core/server/common/src/main/java/alluxio/master/BackupManager.java

```diff
@@ -83,27 +215,130 @@ public void backup(OutputStream os) throws IOException {
    * @param is an input stream to read from the backup
    */
   public void initFromBackup(InputStream is) throws IOException {
-    int count = 0;
     try (GzipCompressorInputStream gzIn = new GzipCompressorInputStream(is);
-         JournalEntryStreamReader reader = new JournalEntryStreamReader(gzIn)) {
+        JournalEntryStreamReader reader = new JournalEntryStreamReader(gzIn)) {
       List<Master> masters = mRegistry.getServers();
-      JournalEntry entry;
+
+      // Executor for reading backup content.
+      ExecutorService backupReaderExecutor =
+          Executors.newSingleThreadExecutor(ThreadFactoryUtils.build("backup-reader-%d", true));
+      // Executor for applying backup.
+      ExecutorService backupApplierExecutor =
+          Executors.newSingleThreadExecutor(ThreadFactoryUtils.build("backup-applier-%d", true));
+
+      // Entry queue will be used as a buffer and synchronization between readers and appliers.
+      LinkedBlockingQueue<JournalEntry> journalEntryQueue = new LinkedBlockingQueue<>(
+          ServerConfiguration.getInt(PropertyKey.MASTER_BACKUP_ENTRY_BUFFER_COUNT));
+
+      // Futures for reader/writer tasks.
+      SettableFuture<Void> readerTaskFuture = SettableFuture.create();
+      SettableFuture<Void> applierTaskFuture = SettableFuture.create();
+
+      // Index masters by name.
       Map<String, Master> mastersByName = Maps.uniqueIndex(masters, Master::getName);
-      while ((entry = reader.readEntry()) != null) {
-        String masterName = JournalEntryAssociation.getMasterForEntry(entry);
-        Master master = mastersByName.get(masterName);
+
+      // Pre-create journal contexts.
+      // They should be closed after applying the backup.
+      Map<Master, JournalContext> masterJCMap = new HashMap<>();
+      for (Master master : masters) {
+        masterJCMap.put(master, master.createJournalContext());
+      }
+
+      // Shows how many entries have been applied.
+      AtomicLong appliedEntryCount = new AtomicLong(0);
+
+      // Create backup reader task.
+      backupReaderExecutor.submit(() -> {
         try {
-          master.processJournalEntry(entry);
-        } catch (Throwable t) {
-          JournalUtils.handleJournalReplayFailure(
-              LOG, t, "Failed to process journal entry %s when init from backup", entry);
+          JournalEntry entry;
+          while (!applierTaskFuture.isDone() && (entry = reader.readEntry()) != null) {
+            // Try to push the entry as long as applier is alive.
+            while (!applierTaskFuture.isDone()) {
```

Comment:
is this needed when it's also checked in the outer while loop?

Reply:
offer() obeys the limits. if an applier is failed, that offer can stuck there forever. That's there is a timeout and checking if if offering a new entry still makes sense. In short, first one is for iterating the entries from backup, the second is for pushing an individual one to buffer.

Reply:
got it, makes sense